ethrpc2/block.go

```diff
@@ -0,0 +1,145 @@
+package ethrpc2
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/0xsequence/ethkit/go-ethereum"
+	"github.com/0xsequence/ethkit/go-ethereum/common"
+	"github.com/0xsequence/ethkit/go-ethereum/common/hexutil"
+	"github.com/0xsequence/ethkit/go-ethereum/core/types"
+)
+
+type rpcTransaction struct {
+	tx *types.Transaction
+	txExtraInfo
+}
+
+type txExtraInfo struct {
+	BlockNumber *string         `json:"blockNumber,omitempty"`
+	BlockHash   *common.Hash    `json:"blockHash,omitempty"`
+	From        *common.Address `json:"from,omitempty"`
+	TxType      string          `json:"type,omitempty"`
+}
+
+type rpcBlock struct {
+	Hash         common.Hash      `json:"hash"`
+	Transactions []rpcTransaction `json:"transactions"`
+	UncleHashes  []common.Hash    `json:"uncles"`
+}
+
+func intoBlock(raw json.RawMessage, ret **types.Block) error {
+	if len(raw) == 0 {
+		return ethereum.NotFound
+	}
+
+	// Decode header and transactions
+	var (
+		head *types.Header
+		body rpcBlock
+	)
+	if err := json.Unmarshal(raw, &head); err != nil {
+		return err
+	}
+	if head == nil {
+		return ethereum.NotFound
+	}
+	if err := json.Unmarshal(raw, &body); err != nil {
+		return err
+	}
+
+	// Fill the sender cache of transactions in the block.
+	txs := make([]*types.Transaction, 0, len(body.Transactions))
+	for _, tx := range body.Transactions {
+		if tx.From != nil {
+			setSenderFromServer(tx.tx, *tx.From, body.Hash)
+		}
+
+		if tx.txExtraInfo.TxType != "" {
+			txType, err := hexutil.DecodeUint64(tx.txExtraInfo.TxType)
+			if err != nil {
+				return err
+			}
+			if txType > types.DynamicFeeTxType {
+				// skip the txn, its a non-standard type we don't care about
+				continue
+			}
+		}
+
+		txs = append(txs, tx.tx)
+	}
+
+	// return types.NewBlockWithHeader(head).WithBody(txs, uncles), nil
+	block := types.NewBlockWithHeader(head).WithBody(txs, nil)
+
+	// TODO: Remove this, we shouldn't need to use the block cache
+	// in order for it to contain the correct block hash
+	block.SetHash(body.Hash)
+
+	*ret = block
+	return nil
+}
+
+func intoTransaction(raw json.RawMessage, tx **types.Transaction) error {
+	return intoTransactionWithPending(raw, tx, nil)
+}
+
+func intoTransactionWithPending(raw json.RawMessage, tx **types.Transaction, pending *bool) error {
```

Comment:
double pointer...? does it feel fine to you..? I'm not sure about performance, or the need for double indirection -- but maybe its fine? lmk